tuf/api/metadata.py

```diff
@@ -0,0 +1,183 @@
+                        break
+
+        return len(verified_keyids) >= self.keyring.threshold.min
+
+    def write_to_json(self, filename: str) -> None:
+        with open(filename, 'r+b') as f:
+            f.write(_get_written_metadata(self.sign()))
+
+class Timestamp(Metadata):
+    def __init__(self, consistent_snapshot: bool = True, expiration: relativedelta = relativedelta(days=1), keyring: KeyRing = None, version: int = 1):
+        super().__init__(consistent_snapshot, expiration, keyring, version)
+
+    def signable(self):
+        expires = self.expiration.replace(tzinfo=None).isoformat()+'Z'
+        filedict = self.signed['meta']
+        return tuf.formats.build_dict_conforming_to_schema(
+            tuf.formats.TIMESTAMP_SCHEMA, version=self.version,
+            expires=expires, meta=filedict)
+
+    # Update metadata about the snapshot metadata.
+    def update(self, version: int, length: int, hashes: JsonDict):
+        fileinfo = self.signed['meta']['snapshot.json']
+        fileinfo['version'] = version
+        fileinfo['length'] = length
+        fileinfo['hashes'] = hashes
+
+class Snapshot(Metadata):
+    def __init__(self, consistent_snapshot: bool = True, expiration: relativedelta = relativedelta(days=1), keyring: KeyRing = None, version: int = 1):
+        super().__init__(consistent_snapshot, expiration, keyring, version)
+        self.targets_fileinfo = {}
+
+    def read_from_json(self, filename: str) -> None:
+        super().read_from_json(filename)
+        meta = self.signed['meta']
+        for target_role in meta:
+            version = meta[target_role]['version']
+            length = meta[target_role].get('length')
+            hashes = meta[target_role].get('hashes')
+            self.targets_fileinfo[target_role] = tuf.formats.make_metadata_fileinfo(version, length, hashes)
+
+    def signable(self):
+        # TODO: probably want to generalise this, a @property.getter in Metadata?
+        expires = self.expiration.replace(tzinfo=None).isoformat()+'Z'
+        return tuf.formats.build_dict_conforming_to_schema(
+            tuf.formats.SNAPSHOT_SCHEMA, version=self.version,
+            expires=expires, meta=self.targets_fileinfo)
+
+    # Add or update metadata about the targets metadata.
+    def update(self, rolename: str, version: int, length: Optional[int] = None, hashes: Optional[JsonDict] = None):
+        self.targets_fileinfo[f'{rolename}.json'] = tuf.formats.make_metadata_fileinfo(version, length, hashes)
+
+class Targets(Metadata):
+    def __init__(self, consistent_snapshot: bool = True, expiration: relativedelta = relativedelta(days=1), keyring: KeyRing = None, version: int = 1):
+        super().__init__(consistent_snapshot, expiration, relativedelta, keyring, version)
+
+    # FIXME
+    def signable(self):
+        return generate_targets_metadata()
+
+    # Add or update metadata about the target.
+    # TODO: how to handle writing consistent targets?
```

Comment:
From the initial API design I assume that it covers only the case of using existing fileinfo. If this is true then writing consistent targets is not applicable here:
https://github.com/theupdateframework/tuf/blob/de8649e12a850fccbc745f0129d43956ef7da33b/tuf/repository_lib.py#L155

If we want to add the option to NOT use existing fileinfo and using tuf for calculating hashes, we need to extend the api (maybe on a later pass) with targets paths and storage_backend.